### PR TITLE
Screenshot support for Supermicros

### PR DESCRIPTION
**Pull request by joelrebel**
## What does this PR implement/change/remove?

- Implements the Supermicro provider with screen capture support.
- Adds an error type to exclude incompatible provider related errors
- Updates the Dell provider to check for device manufacturer.

### Checklist
- [X] Tests added
- [X] Similar commits squashed

### The HW vendor this change applies to (if applicable)

Supermicro

### The HW model number, product name this change applies to (if applicable)

SYS-5019C-MR
SYS-510T-MR
6029P-E1CR12L

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -32,14 +33,15 @@ type Client struct {
...
-	oneTimeRegistryEnabled bool
-	providerConfig         providerConfig
+	httpClient                       *http.Client
+	httpClientSetupFuncs             []func(*http.Client)
+	mdLock                           *sync.Mutex
+	metadata                         *bmc.Metadata
+	perProviderTimeout               func(context.Context) time.Duration
+	oneTimeRegistry                  *registrar.Registry
+	oneTimeRegistryEnabled           bool
+	returnIncompatibleProviderErrors bool
```
Comment (jacobweinstock) [inline client.go:43]: Hey @joelrebel, I like this idea! I had a similar thought a little while back. I created a branch for it but forgot to create the PR. I just opened the PR now. https://github.com/bmc-toolbox/bmclib/pull/335 It's a bit different approach. Let me know what you think. Thanks!

File: providers/supermicro/supermicro.go [hunk 2]
```diff
@@ -0,0 +1,343 @@
+package supermicro
```
Comment (jacobweinstock) [inline providers/supermicro/supermicro.go:1]: There are a few code comments referencing `redfish` in this file. Might want to be `vendorapi`? Maybe copy/paste left over?

**Comment by jacobweinstock** [comment]
I can also confirm that this works on the `SYS-E300-9D-4CN8TP`!

File: providers/supermicro/supermicro.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+)
+
+var (
+	// Features implemented
+	Features = registrar.Features{
+		providers.FeatureScreenshot,
+	}
+)
+
+// supports
```
Comment (jacobweinstock) [inline providers/supermicro/supermicro.go:43]: I can also confirm that this works on the SYS-E300-9D-4CN8TP!

File: client.go (hunk 1 above)
Comment (joelrebel) [reply client.go:43]: Hey @jacobweinstock thanks for taking a look, your approach is better in this case and we can include that instead. I'll remove the parts that overlap in my PR.

File: providers/supermicro/supermicro.go (hunk 2 above)
Comment (joelrebel) [reply providers/supermicro/supermicro.go:1]: oh, right, fixed!

File: client.go (hunk 1 above)
Comment (joelrebel) [reply client.go]: updated in latest commits

File: providers/supermicro/supermicro.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+	host      string
+	user      string
+	pass      string
+	port      string
+	csrfToken string
+	log       logr.Logger
+}
+
+// New returns connection with a Supermicro client initialized
+func NewClient(host, user, pass string, log logr.Logger, opts ...Option) *Client {
```
Comment (jacobweinstock) [inline providers/supermicro/supermicro.go:88]: might think about making `log` optional a then default to `logr.Discard`. Not a PR blocker.

File: providers/supermicro/supermicro.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+	if status != 200 {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, strconv.Itoa(status))
+	}
+
+	if !bytes.Contains(body, []byte(`url_redirect.cgi?url_name=mainmenu`)) &&
+		!bytes.Contains(body, []byte(`url_redirect.cgi?url_name=topmenu`)) {
+		return errors.Wrap(bmclibErrs.ErrLoginFailed, "unexpected response contents")
+	}
+
+	contentsTopMenu, status, err := c.query(ctx, "cgi/url_redirect.cgi?url_name=topmenu", http.MethodGet, nil, nil, 0)
```
Comment (jacobweinstock) [inline providers/supermicro/supermicro.go:135]: might be helpful to have a code comment about needing to first POST to login and then do a GET for the csrf token. Seems like something i'm going to forget pretty quickly 😬

File: providers/supermicro/supermicro.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+func (c *Client) Screenshot(ctx context.Context) (image []byte, fileType string, err error) {
+	fileType = "jpg"
+
+	// request screen preview to be saved
+	if err := c.initScreenPreview(ctx); err != nil {
+		return nil, "", err
+	}
+
+	// give the bmc a few seconds to store the screen preview
+	time.Sleep(2 * time.Second)
```
Comment (jacobweinstock) [inline providers/supermicro/supermicro.go:213]: ah...this is unfortunate. I imagine this is the best we can do? there's no way programmatically to determine whether the screen preview was stored?

**Review by jacobweinstock — APPROVED** [review]
left a few comments, nothing blocking. Thanks @joelrebel. This is a cool feature. I can see it being very useful in a few different ways!

File: providers/supermicro/supermicro.go (hunk 6 above)
Comment (joelrebel) [reply providers/supermicro/supermicro.go:213]: Yeah, theres no means to determine if the is ready. The SMC web interface just delays the fetch for a few seconds.

File: providers/supermicro/supermicro.go (hunk 5 above)
Comment (joelrebel) [reply providers/supermicro/supermicro.go:135]: noted, will include that in the next set of changes

**Pull request merged**